# You are what you eat - Relating Demographic Data to Food Consumption Habits

In [28]:
#%matplotlib
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns  # needed for heatmap
from os.path import join
plt.rcParams['figure.figsize'] = [7, 5]

In [29]:
# define paths for data
FOLDER_PATH = 'data'
PURCHASES_PATH = 'area_level_purchases'
OSWARD_GROCERY = 'year_osward_grocery.csv'
WARD_ATLAS = 'ward-atlas-data.csv'

ward_data_path = join(FOLDER_PATH, PURCHASES_PATH, OSWARD_GROCERY)
atlas_data_path = join(FOLDER_PATH, WARD_ATLAS)

In [30]:
# load the data sets
df_ward = pd.read_csv(ward_data_path, sep=',', index_col=0)
df_atlas = pd.read_csv(atlas_data_path, sep=',', encoding='iso-8859-1')

## Atlas dataset preprocessing

In [31]:
df_atlas

Unnamed: 0 Unnamed: 1            Unnamed: 2                   Unnamed: 3  \
0        Codes   New Code               Borough                        Names   
1         00AA  E09000001        City of London               City of London   
2       00ABFX  E05000026  Barking and Dagenham                        Abbey   
3       00ABFY  E05000027  Barking and Dagenham                       Alibon   
4       00ABFZ  E05000028  Barking and Dagenham                    Becontree   
5       00ABGA  E05000029  Barking and Dagenham               Chadwell Heath   
6       00ABGB  E05000030  Barking and Dagenham                    Eastbrook   
7       00ABGC  E05000031  Barking and Dagenham                     Eastbury   
8       00ABGD  E05000032  Barking and Dagenham                    Gascoigne   
9       00ABGE  E05000033  Barking and Dagenham                   Goresbrook   
10      00ABGF  E05000034  Barking and Dagenham                        Heath   
11      00ABGG  E05000035  Barking and Dagenham                   Longbridge   
12      00ABGH  E05000036  Barking and Dagenham                   Mayesbrook   
13      00ABGJ  E05000037  Barking and Dagenham                     Parsloes   
14      00ABGK  E05000038  Barking and Dagenham                        River   
15      00ABGL  E05000039  Barking and Dagenham                       Thames   
16      00ABGM  E05000040  Barking and Dagenham                      Valence   
17      00ABGN  E05000041  Barking and Dagenham                      Village   
18      00ABGP  E05000042  Barking and Dagenham                    Whalebone   
19      00ACFX  E05000043                Barnet               Brunswick Park   
20      00ACFY  E05000044                Barnet                    Burnt Oak   
21      00ACFZ  E05000045                Barnet                  Childs Hill   
22      00ACGA  E05000046                Barnet                    Colindale   
23      00ACGB  E05000047                Barnet                     Coppetts   
24      00ACGC  E05000048                Barnet                  East Barnet   
25      00ACGD  E05000049                Barnet                East Finchley   
26      00ACGE  E05000050                Barnet                      Edgware   
27      00ACGF  E05000051                Barnet          Finchley Church End   
28      00ACGG  E05000052                Barnet                Garden Suburb   
29      00ACGH  E05000053                Barnet                Golders Green   
..         ...        ...                   ...                          ...   
599     00BJGN  E05000623            Wandsworth                  Shaftesbury   
600     00BJGP  E05000624            Wandsworth                  Southfields   
601     00BJGQ  E05000625            Wandsworth                  Thamesfield   
602     00BJGR  E05000626            Wandsworth                      Tooting   
603     00BJGS  E05000627            Wandsworth            Wandsworth Common   
604     00BJGT  E05000628            Wandsworth                    West Hill   
605     00BJGU  E05000629            Wandsworth                  West Putney   
606     00BKGA  E05000630           Westminster                   Abbey Road   
607     00BKGB  E05000631           Westminster                    Bayswater   
608     00BKGC  E05000632           Westminster  Bryanston and Dorset Square   
609     00BKGD  E05000633           Westminster                    Churchill   
610     00BKGE  E05000634           Westminster                Church Street   
611     00BKGF  E05000635           Westminster                  Harrow Road   
612     00BKGG  E05000636           Westminster                    Hyde Park   
613     00BKGH  E05000637           Westminster  Knightsbridge and Belgravia   
614     00BKGJ  E05000638           Westminster               Lancaster Gate   
615     00BKGK  E05000639           Westminster                Little Venice   
616     00BKGL  E05000640           Westminster                   Maida Vale   
617 

In [32]:
df_atlas.columns.values

array(['Unnamed: 0', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3',
       'Population and Age; Population Estimates and Projections; 2001',
       'Population and Age; Population Estimates and Projections; 2006',
       'Population and Age; Population Estimates and Projections; 2011',
       'Population and Age; Population Estimates and Projections; 2013',
       'Population and Age; Population Estimates and Projections; 2014',
       'Population and Age; Population Estimates and Projections; 2015',
       'Population and Age; Population Estimates and Projections; 2016',
       'Population and Age; Population Estimates and Projections; 2018',
       'Population and Age; Population Estimates and Projections; 2023',
       'Population and Age; Population Estimates and Projections; 2028',
       'Population and Age; Age structure (percentage) - 2013; Aged 0-15',
       'Population and Age; Age structure (percentage) - 2013; Aged 16-64',
       'Population and Age; Age structure (percentage) -

In [33]:
print(f"Rows: {df_atlas.shape[0]}, Columns: {df_atlas.shape[1]} of df_atlas")

Rows: 629, Columns: 946 of df_atlas


The `df_atlas` data set looks very messy. Row 0 contains info about the data. For every ward and demographic marker yearly (from 2001 to 2006) values exist in columns. According to the [source](https://data.london.gov.uk/dataset/ward-profiles-and-atlas) this data set was collected in September 2015 with the most up-to-date information available. Thus, for many markers, the last available information dates to 2015 or even earlier (e.g. 2013). Our first step is to clean the data set from outdated columns and only keep the most up-to-date demographic markers.

As indicated by the title, we are interested in demographic markers which help classifying different classes of people. These are for example gender, religion, ethnicity, wealth, age. Thus, in the following, we drop columns which we deem not helpful in classifying people into groups. We acknowledge that this process can be rather biased, however, as the data set has 946 columns, we see the need to reduce the data set's complexity.

In [34]:
# clean age (columns after 2013 are projections, therefore keep 2013 columns)
cond = df_atlas.columns.str.startswith('Population and Age') & ~df_atlas.columns.str.contains('2013')
df_atlas = df_atlas.loc[:, ~cond]
# drop absolute population values
df_atlas = df_atlas.drop(['Population and Age; Population Estimates and Projections; 2013',
               'Population and Age; Aged 0-15; 2013',
               'Population and Age; Aged 16-64; 2013',
               'Population and Age; Aged 65+; 2013',
               'Population and Age; All ages; 2013',
               'Population and Age; Mean age; 2013',
               'Population and Age; Median age; 2013'], axis=1)
# delete population density
cond = df_atlas.columns.str.startswith('Area and Density')
df_atlas = df_atlas.loc[:, ~cond]
# delete the 18 ethnic groups for now, as they add complexity
cond = df_atlas.columns.str.startswith('Diversity; Ethnic Group 18 groups')
df_atlas = df_atlas.loc[:, ~cond]
# delete household language (of no interest) 
cond = df_atlas.columns.str.startswith('Diversity; Household Language - 2011 Census')
df_atlas = df_atlas.loc[:, ~cond]
# delete household composition data
cond = df_atlas.columns.str.startswith('Household composition')
df_atlas = df_atlas.loc[:, ~cond]
# delete household repossesion
cond = df_atlas.columns.str.startswith('Home repossessions')
df_atlas = df_atlas.loc[:, ~cond]
# delete outdated birth and death markers
cond = df_atlas.columns.str.startswith('Births and deaths')
df_atlas = df_atlas.loc[:, ~cond]
# delete outdated life expectancy
cond = df_atlas.columns.str.startswith('Life Expectancy')
df_atlas = df_atlas.loc[:, ~cond]
# delete outdated housing markers
cond = df_atlas.columns.str.startswith('Housing type and Tenure')
df_atlas = df_atlas.loc[:, ~cond]
cond = df_atlas.columns.str.startswith('House Prices')
df_atlas = df_atlas.loc[:, ~cond]
# delete tax related data (not of interest)
cond = df_atlas.columns.str.startswith('Dwellings and Council tax')
df_atlas = df_atlas.loc[:, ~cond]
# delete dwelling data which is not of interest
cond = df_atlas.columns.str.startswith('Property Type') | df_atlas.columns.str.startswith('Size of dwellings') | df_atlas.columns.str.startswith('Property build period')
df_atlas = df_atlas.loc[:, ~cond]
# delete outdated income data
cond = df_atlas.columns.str.startswith('Household Income') & ~df_atlas.columns.str.contains('2012/13')
df_atlas = df_atlas.loc[:, ~cond]
# outdated employment data
cond = df_atlas.columns.str.startswith('Employment')
df_atlas = df_atlas.loc[:, ~cond]
# delete benefits claimants and jobseekers allowance
cond = df_atlas.columns.str.startswith('Benefits claimants') | df_atlas.columns.str.startswith('Jobseekers Allowance')
df_atlas = df_atlas.loc[:, ~cond]
# delete out of work families
cond = df_atlas.columns.str.startswith('Out-of-Work Families')
df_atlas = df_atlas.loc[:, ~cond]
# outdated poverty
cond = df_atlas.columns.str.startswith('Children in Poverty')
df_atlas = df_atlas.loc[:, ~cond]
# outdated "intelligence" scores
cond = df_atlas.columns.str.startswith('GCSE')
df_atlas = df_atlas.loc[:, ~cond]
cond = df_atlas.columns.str.startswith('A-Level Point Scores')
df_atlas = df_atlas.loc[:, ~cond]
# pupil abscence not if interest
cond = df_atlas.columns.str.startswith('Pupil Absence')
df_atlas = df_atlas.loc[:, ~cond]
# individual crimes not of interest, only crime rate
cond = df_atlas.columns.str.startswith('Crime') # & ~df_atlas.columns.str.contains('Total crime rate; 2014/15')
df_atlas = df_atlas.loc[:, ~cond]
# further columns not of interest
cond = df_atlas.columns.str.startswith('Fires') | df_atlas.columns.str.startswith('Ambulance') \
| df_atlas.columns.str.startswith('Binge Drinking') | df_atlas.columns.str.startswith('Road Casualties') \
| df_atlas.columns.str.startswith('Air Emissions') | df_atlas.columns.str.startswith('Land Use') \
| df_atlas.columns.str.startswith('Access to green space and nature') | df_atlas.columns.str.startswith('Public Transport Accessibility')\
| df_atlas.columns.str.startswith('Car access') | df_atlas.columns.str.startswith('Travel to work by bicycle') \
| df_atlas.columns.str.startswith('Workplace employment') | df_atlas.columns.str.startswith('Nat Insurance No. registrations') \
| df_atlas.columns.str.startswith('Election turnout')
df_atlas = df_atlas.loc[:, ~cond]
# obesity already available in cleaned format
cond = df_atlas.columns.str.startswith('Childhood Obesity Prevalence')
df_atlas = df_atlas.loc[:, ~cond]
# delete old indices of deprivation
cond = df_atlas.columns.str.startswith('Indices of Deprivation') & ~df_atlas.columns.str.contains('2010')
df_atlas = df_atlas.loc[:, ~cond]

The indices of deprivation measure how deprived an area is [source](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/464597/English_Indices_of_Deprivation_2015_-_Research_Report.pdf). In essence, these indicators indicate how impoverished an area is. The linked documents lists the meaning of the different indices. *The average score measure summarises the average level of deprivation across the higher-level area,based on the scores of the Lower-layer Super Output Areas in the area.* Therefore, we decided to keep only this measure, as we are interested in an indicator of "average impoverishment".

In [35]:
cond = df_atlas.columns.str.startswith('Indices of Deprivation') & ~df_atlas.columns.str.endswith('Average Score; 2010')
df_atlas = df_atlas.loc[:, ~cond]

In [36]:
df_atlas.columns.values

array(['Unnamed: 0', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3',
       'Population and Age; Age structure (percentage) - 2013; Aged 0-15',
       'Population and Age; Age structure (percentage) - 2013; Aged 16-64',
       'Population and Age; Age structure (percentage) - 2013; Aged 65+',
       'Diversity; Ethnic Group 5 groups - 2011 Census; White',
       'Diversity; Ethnic Group 5 groups - 2011 Census; Mixed',
       'Diversity; Ethnic Group 5 groups - 2011 Census; Asian or Asian British',
       'Diversity; Ethnic Group 5 groups - 2011 Census; Black or Black British',
       'Diversity; Ethnic Group 5 groups - 2011 Census; Other',
       'Diversity; Religion - 2011 Census; Christian',
       'Diversity; Religion - 2011 Census; Buddhist',
       'Diversity; Religion - 2011 Census; Hindu',
       'Diversity; Religion - 2011 Census; Jewish',
       'Diversity; Religion - 2011 Census; Muslim',
       'Diversity; Religion - 2011 Census; Sikh',
       'Diversity; Religion - 2011 Census; O

In [37]:
print(f"Rows: {df_atlas.shape[0]}, Columns: {df_atlas.shape[1]} of df_atlas")

Rows: 629, Columns: 37 of df_atlas


In [38]:
df_atlas.head(3)

Unnamed: 0 Unnamed: 1            Unnamed: 2      Unnamed: 3  \
0      Codes   New Code               Borough           Names   
1       00AA  E09000001        City of London  City of London   
2     00ABFX  E05000026  Barking and Dagenham           Abbey   

  Population and Age; Age structure (percentage) - 2013; Aged 0-15  \
0                                          Aged 0-15                 
1                                                7.5                 
2                                        25.27472527                 

  Population and Age; Age structure (percentage) - 2013; Aged 16-64  \
0                                         Aged 16-64                  
1                                               77.5                  
2                                        69.96336996                  

  Population and Age; Age structure (percentage) - 2013; Aged 65+  \
0                                           Aged 65+                
1                                                 15                
2                                        5.128205128                

  Diversity; Ethnic Group 5 groups - 2011 Census; White  \
0                                              White      
1                                               5799      
2                                               3596      

  Diversity; Ethnic Group 5 groups - 2011 Census; Mixed  \
0                                              Mixed      
1                                                289      
2                                                585      

  Diversity; Ethnic Group 5 groups - 2011 Census; Asian or Asian British  ...  \
0                             Asian or Asian British                      ...   
1                                                940                      ...   
2                                               5619                      ...   

  Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Level 2 qualifications  \
0  % Highest level of qualification: Level 2 qual...                                                                    
1                                                6.6                                                                    
2                                               11.5                                                                    

  Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Apprenticeship  \
0   % Highest level of qualification: Apprenticeship                                                            
1                                                0.7                                                            
2                                                1.1                                                            

  Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Level 3 qualifications  \
0  % Highest level of qualification: Level 3 qual...                                                                    
1                                                7.2                                                                    
2                                                8.5                                                                    

  Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Level 4 qualifications and above  \
0  % Highest level of qualification: Level 4 qual...                                                                              
1                                               68.4                                                                              
2                                               34.5                                                                              

  Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Other qualifications  \
0  % Highest level of qualificat

In [39]:
# The first 4 columns seem to have a wrong name
df_atlas.rename(columns={'Unnamed: 0': 'Codes', 'Unnamed: 1': 'New Code', 'Unnamed: 2': 'Borough' , 'Unnamed: 3': 'Names'}, inplace=True)

In [40]:
df_atlas

Codes   New Code               Borough                        Names  \
0     Codes   New Code               Borough                        Names   
1      00AA  E09000001        City of London               City of London   
2    00ABFX  E05000026  Barking and Dagenham                        Abbey   
3    00ABFY  E05000027  Barking and Dagenham                       Alibon   
4    00ABFZ  E05000028  Barking and Dagenham                    Becontree   
5    00ABGA  E05000029  Barking and Dagenham               Chadwell Heath   
6    00ABGB  E05000030  Barking and Dagenham                    Eastbrook   
7    00ABGC  E05000031  Barking and Dagenham                     Eastbury   
8    00ABGD  E05000032  Barking and Dagenham                    Gascoigne   
9    00ABGE  E05000033  Barking and Dagenham                   Goresbrook   
10   00ABGF  E05000034  Barking and Dagenham                        Heath   
11   00ABGG  E05000035  Barking and Dagenham                   Longbridge   
12   00ABGH  E05000036  Barking and Dagenham                   Mayesbrook   
13   00ABGJ  E05000037  Barking and Dagenham                     Parsloes   
14   00ABGK  E05000038  Barking and Dagenham                        River   
15   00ABGL  E05000039  Barking and Dagenham                       Thames   
16   00ABGM  E05000040  Barking and Dagenham                      Valence   
17   00ABGN  E05000041  Barking and Dagenham                      Village   
18   00ABGP  E05000042  Barking and Dagenham                    Whalebone   
19   00ACFX  E05000043                Barnet               Brunswick Park   
20   00ACFY  E05000044                Barnet                    Burnt Oak   
21   00ACFZ  E05000045                Barnet                  Childs Hill   
22   00ACGA  E05000046                Barnet                    Colindale   
23   00ACGB  E05000047                Barnet                     Coppetts   
24   00ACGC  E05000048                Barnet                  East Barnet   
25   00ACGD  E05000049                Barnet                East Finchley   
26   00ACGE  E05000050                Barnet                      Edgware   
27   00ACGF  E05000051                Barnet          Finchley Church End   
28   00ACGG  E05000052                Barnet                Garden Suburb   
29   00ACGH  E05000053                Barnet                Golders Green   
..      ...        ...                   ...                          ...   
599  00BJGN  E05000623            Wandsworth                  Shaftesbury   
600  00BJGP  E05000624            Wandsworth                  Southfields   
601  00BJGQ  E05000625            Wandsworth                  Thamesfield   
602  00BJGR  E05000626            Wandsworth                      Tooting   
603  00BJGS  E05000627            Wandsworth            Wandsworth Common   
604  00BJGT  E05000628            Wandsworth                    West Hill   
605  00BJGU  E05000629            Wandsworth                  West Putney   
606  00BKGA  E05000630           Westminster                   Abbey Road   
607  00BKGB  E05000631           Westminster                    Bayswater   
608  00BKGC  E05000632           Westminster  Bryanston and Dorset Square   
609  00BKGD  E05000633           Westminster                    Churchill   
610  00BKGE  E05000634           Westminster                Church Street   
611  00BKGF  E05000635           Westminster                  Harrow Road   
612  00BKGG  E05000636           Westminster                    Hyde Park   
613  00BKGH  E05000637           Westminster  Knightsbridge and Belgravia   
614  00BKGJ  E05000638           Westminster               Lancaster Gate   
615  00BKGK  E05000639           Westminster                Little Venice   
616  00BKGL  E05000640           Westminster                   Maida Vale   
617  00BKGM  E05000641           Westminster       Marylebone High Street   
618  00BKGN  E05000642           Westminster                 Queen's Park   
61

In [41]:
# the last three rows contain the demographic markers on country level and our not needed for our analysis
df_atlas.iloc[-3:]

Codes   New Code Borough              Names  \
626    #1  E12000007     NaN             London   
627    #2  E92000001     NaN            England   
628    #3  K04000001     NaN  England and Wales   

    Population and Age; Age structure (percentage) - 2013; Aged 0-15  \
626                                        19.82387476                 
627                                        18.95309227                 
628                                        18.89198229                 

    Population and Age; Age structure (percentage) - 2013; Aged 16-64  \
626                                        68.81515745                  
627                                        63.77216928                  
628                                        64.06230146                  

    Population and Age; Age structure (percentage) - 2013; Aged 65+  \
626                                        11.36037479                
627                                        17.27473845                
628                                        17.04571626                

    Diversity; Ethnic Group 5 groups - 2011 Census; White  \
626                                            4887435      
627                                           45281142      
628                                           48209395      

    Diversity; Ethnic Group 5 groups - 2011 Census; Mixed  \
626                                             405279      
627                                            1192879      
628                                            1224400      

    Diversity; Ethnic Group 5 groups - 2011 Census; Asian or Asian British  \
626                                            1511546                       
627                                            4143403                       
628                                            4213531                       

     ...  \
626  ...   
627  ...   
628  ...   

    Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Level 2 qualifications  \
626                                               11.8                                                                    
627                                        15.22370749                                                                    
628                                               15.3                                                                    

    Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Apprenticeship  \
626                                                1.6                                                            
627                                        3.565823564                                                            
628                                                3.6                                                            

    Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Level 3 qualifications  \
626                                               10.5                                                                    
627                                        12.35096053                                                                    
628                                               12.3                                                                    

    Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Level 4 qualifications and above  \
626                                               37.7                                                                              
627                                        27.37721571                                                                              
628                                               27.2                                                                              

    Qualifications; Qualifications and Students - 2011 Census; % Highest level of qualification: Oth

In [42]:
# delete the last three rows
df_atlas.drop(df_atlas.index[-3:], inplace=True)
df_atlas

Codes   New Code               Borough                        Names  \
0     Codes   New Code               Borough                        Names   
1      00AA  E09000001        City of London               City of London   
2    00ABFX  E05000026  Barking and Dagenham                        Abbey   
3    00ABFY  E05000027  Barking and Dagenham                       Alibon   
4    00ABFZ  E05000028  Barking and Dagenham                    Becontree   
5    00ABGA  E05000029  Barking and Dagenham               Chadwell Heath   
6    00ABGB  E05000030  Barking and Dagenham                    Eastbrook   
7    00ABGC  E05000031  Barking and Dagenham                     Eastbury   
8    00ABGD  E05000032  Barking and Dagenham                    Gascoigne   
9    00ABGE  E05000033  Barking and Dagenham                   Goresbrook   
10   00ABGF  E05000034  Barking and Dagenham                        Heath   
11   00ABGG  E05000035  Barking and Dagenham                   Longbridge   
12   00ABGH  E05000036  Barking and Dagenham                   Mayesbrook   
13   00ABGJ  E05000037  Barking and Dagenham                     Parsloes   
14   00ABGK  E05000038  Barking and Dagenham                        River   
15   00ABGL  E05000039  Barking and Dagenham                       Thames   
16   00ABGM  E05000040  Barking and Dagenham                      Valence   
17   00ABGN  E05000041  Barking and Dagenham                      Village   
18   00ABGP  E05000042  Barking and Dagenham                    Whalebone   
19   00ACFX  E05000043                Barnet               Brunswick Park   
20   00ACFY  E05000044                Barnet                    Burnt Oak   
21   00ACFZ  E05000045                Barnet                  Childs Hill   
22   00ACGA  E05000046                Barnet                    Colindale   
23   00ACGB  E05000047                Barnet                     Coppetts   
24   00ACGC  E05000048                Barnet                  East Barnet   
25   00ACGD  E05000049                Barnet                East Finchley   
26   00ACGE  E05000050                Barnet                      Edgware   
27   00ACGF  E05000051                Barnet          Finchley Church End   
28   00ACGG  E05000052                Barnet                Garden Suburb   
29   00ACGH  E05000053                Barnet                Golders Green   
..      ...        ...                   ...                          ...   
596  00BJGK  E05000620            Wandsworth                   Queenstown   
597  00BJGL  E05000621            Wandsworth                   Roehampton   
598  00BJGM  E05000622            Wandsworth              St. Mary's Park   
599  00BJGN  E05000623            Wandsworth                  Shaftesbury   
600  00BJGP  E05000624            Wandsworth                  Southfields   
601  00BJGQ  E05000625            Wandsworth                  Thamesfield   
602  00BJGR  E05000626            Wandsworth                      Tooting   
603  00BJGS  E05000627            Wandsworth            Wandsworth Common   
604  00BJGT  E05000628            Wandsworth                    West Hill   
605  00BJGU  E05000629            Wandsworth                  West Putney   
606  00BKGA  E05000630           Westminster                   Abbey Road   
607  00BKGB  E05000631           Westminster                    Bayswater   
608  00BKGC  E05000632           Westminster  Bryanston and Dorset Square   
609  00BKGD  E05000633           Westminster                    Churchill   
610  00BKGE  E05000634           Westminster                Church Street   
611  00BKGF  E05000635           Westminster                  Harrow Road   
612  00BKGG  E05000636           Westminster                    Hyde Park   
613  00BKGH  E05000637           Westminster  Knightsbridge and Belgravia   
614  00BKGJ  E05000638           Westminster               Lancaster Gate   
615  00BKGK  E05000639           Westminster                Little Venice   
61

In [43]:
# row 0 od df_atlas seems to have column names and years in it, therefore, drop it
df_atlas.drop(df_atlas.index[0], inplace=True)
df_atlas

Codes   New Code               Borough                        Names  \
1      00AA  E09000001        City of London               City of London   
2    00ABFX  E05000026  Barking and Dagenham                        Abbey   
3    00ABFY  E05000027  Barking and Dagenham                       Alibon   
4    00ABFZ  E05000028  Barking and Dagenham                    Becontree   
5    00ABGA  E05000029  Barking and Dagenham               Chadwell Heath   
6    00ABGB  E05000030  Barking and Dagenham                    Eastbrook   
7    00ABGC  E05000031  Barking and Dagenham                     Eastbury   
8    00ABGD  E05000032  Barking and Dagenham                    Gascoigne   
9    00ABGE  E05000033  Barking and Dagenham                   Goresbrook   
10   00ABGF  E05000034  Barking and Dagenham                        Heath   
11   00ABGG  E05000035  Barking and Dagenham                   Longbridge   
12   00ABGH  E05000036  Barking and Dagenham                   Mayesbrook   
13   00ABGJ  E05000037  Barking and Dagenham                     Parsloes   
14   00ABGK  E05000038  Barking and Dagenham                        River   
15   00ABGL  E05000039  Barking and Dagenham                       Thames   
16   00ABGM  E05000040  Barking and Dagenham                      Valence   
17   00ABGN  E05000041  Barking and Dagenham                      Village   
18   00ABGP  E05000042  Barking and Dagenham                    Whalebone   
19   00ACFX  E05000043                Barnet               Brunswick Park   
20   00ACFY  E05000044                Barnet                    Burnt Oak   
21   00ACFZ  E05000045                Barnet                  Childs Hill   
22   00ACGA  E05000046                Barnet                    Colindale   
23   00ACGB  E05000047                Barnet                     Coppetts   
24   00ACGC  E05000048                Barnet                  East Barnet   
25   00ACGD  E05000049                Barnet                East Finchley   
26   00ACGE  E05000050                Barnet                      Edgware   
27   00ACGF  E05000051                Barnet          Finchley Church End   
28   00ACGG  E05000052                Barnet                Garden Suburb   
29   00ACGH  E05000053                Barnet                Golders Green   
30   00ACGJ  E05000054                Barnet                         Hale   
..      ...        ...                   ...                          ...   
596  00BJGK  E05000620            Wandsworth                   Queenstown   
597  00BJGL  E05000621            Wandsworth                   Roehampton   
598  00BJGM  E05000622            Wandsworth              St. Mary's Park   
599  00BJGN  E05000623            Wandsworth                  Shaftesbury   
600  00BJGP  E05000624            Wandsworth                  Southfields   
601  00BJGQ  E05000625            Wandsworth                  Thamesfield   
602  00BJGR  E05000626            Wandsworth                      Tooting   
603  00BJGS  E05000627            Wandsworth            Wandsworth Common   
604  00BJGT  E05000628            Wandsworth                    West Hill   
605  00BJGU  E05000629            Wandsworth                  West Putney   
606  00BKGA  E05000630           Westminster                   Abbey Road   
607  00BKGB  E05000631           Westminster                    Bayswater   
608  00BKGC  E05000632           Westminster  Bryanston and Dorset Square   
609  00BKGD  E05000633           Westminster                    Churchill   
610  00BKGE  E05000634           Westminster                Church Street   
611  00BKGF  E05000635           Westminster                  Harrow Road   
612  00BKGG  E05000636           Westminster                    Hyde Park   
613  00BKGH  E05000637           Westminster  Knightsbridge and Belgravia   
614  00BKGJ  E05000638           Westminster               Lancaster Gate   
615  00BKGK  E05000639           Westminster                Little Venice   
61

In [44]:
# can delete column `Codes` since this represents old area codes, the tesco data set is indexed by `New Code`
df_atlas = df_atlas.drop(['Codes'], axis=1)

In [45]:
df_atlas.columns.values

array(['New Code', 'Borough', 'Names',
       'Population and Age; Age structure (percentage) - 2013; Aged 0-15',
       'Population and Age; Age structure (percentage) - 2013; Aged 16-64',
       'Population and Age; Age structure (percentage) - 2013; Aged 65+',
       'Diversity; Ethnic Group 5 groups - 2011 Census; White',
       'Diversity; Ethnic Group 5 groups - 2011 Census; Mixed',
       'Diversity; Ethnic Group 5 groups - 2011 Census; Asian or Asian British',
       'Diversity; Ethnic Group 5 groups - 2011 Census; Black or Black British',
       'Diversity; Ethnic Group 5 groups - 2011 Census; Other',
       'Diversity; Religion - 2011 Census; Christian',
       'Diversity; Religion - 2011 Census; Buddhist',
       'Diversity; Religion - 2011 Census; Hindu',
       'Diversity; Religion - 2011 Census; Jewish',
       'Diversity; Religion - 2011 Census; Muslim',
       'Diversity; Religion - 2011 Census; Sikh',
       'Diversity; Religion - 2011 Census; Other religions',
       

In [46]:
# very long column names, rename columns to more concise names
import re
rename_map = {name: re.sub('.*; ', '', name) for name in df_atlas.columns.values}
# after inspection 'Household Income; Median Modelled Household income (£); 2012/13',
# 'Household Income; Mean Modelled Household income (£); 2012/13',
# 'Happiness and Well-being; Subjective well-being average score, 2011/12 and 2012/13; ',
# and 'Indices of Deprivation; Average Score; 2010' have been replaced too much, manually add those column names
rename_map['Household Income; Median Modelled Household income (£); 2012/13'] = 'Household Income Mean'
rename_map['Household Income; Mean Modelled Household income (£); 2012/13'] = 'Household Income Median'
rename_map['Happiness and Well-being; Subjective well-being average score, 2011/12 and 2012/13; '] = 'Well-Being'
rename_map['Indices of Deprivation; Average Score; 2010'] = 'IOD AVG'
df_atlas.rename(columns=rename_map, inplace=True)
                         
df_atlas.rename(columns={'% No qualifications': 'No qualifications', 
                         '% Highest level of qualification: Level 1 qualifications': 'Level 1 qualifications', 
                         '% Highest level of qualification: Level 2 qualifications': 'Level 2 qualifications', 
                         '% Highest level of qualification: Apprenticeship': 'Apprenticeship qualifications', 
                         '% Highest level of qualification: Level 3 qualifications': 'Level 3 qualifications', 
                         '% Highest level of qualification: Level 4 qualifications and above': 'Level >=4 qualifications',
                         '% Highest level of qualification: Other qualifications': 'Other qualifications', 
                         '% People with Bad or Very Bad Health': 'Bad Health',
                         '% People whose Day-to-day activities are limited a lot': 'Limited activities'},
                inplace=True)

In [47]:
# checking column type
df_atlas.dtypes

New Code                                                                    object
Borough                                                                     object
Names                                                                       object
Aged 0-15                                                                   object
Aged 16-64                                                                  object
Aged 65+                                                                    object
White                                                                       object
Mixed                                                                       object
Asian or Asian British                                                      object
Black or Black British                                                      object
Other                                                                       object
Christian                                                                   object
Budd

The values in the columns seem to be converted to objects instead of floats or ints, therefore convert these objects to numveric values where possible. None of these columns store string data (except the first three columns).

In [48]:
# to numeric conversion
df_atlas.loc[:, 'Aged 0-15':] = df_atlas.loc[:, 'Aged 0-15':].apply(pd.to_numeric, errors='coerce')
df_atlas.dtypes

New Code                                                                    object
Borough                                                                     object
Names                                                                       object
Aged 0-15                                                                  float64
Aged 16-64                                                                 float64
Aged 65+                                                                   float64
White                                                                        int64
Mixed                                                                        int64
Asian or Asian British                                                       int64
Black or Black British                                                       int64
Other                                                                        int64
Christian                                                                    int64
Budd

In [49]:
# checking for NA values
np.sum(df_atlas.isna())

New Code                                                                   0
Borough                                                                    0
Names                                                                      0
Aged 0-15                                                                  0
Aged 16-64                                                                 0
Aged 65+                                                                   0
White                                                                      0
Mixed                                                                      0
Asian or Asian British                                                     0
Black or Black British                                                     0
Other                                                                      0
Christian                                                                  0
Buddhist                                                                   0

In [50]:
age_cols = ['Aged 0-15', 'Aged 16-64', 'Aged 65+'] #sum to 1
ethnicities_cols = ['White', 'Mixed', 'Asian or Asian British', 'Black or Black British', 'Other'] #sum to 1
religion_cols = ['Christian', 'Buddhist', 'Hindu', 'Jewish', 'Muslim', 'Sikh',
       'No religion', 'Other religions', 'Religion not stated'] # sum to 1
born_cols = ['Born in UK', 'Not Born in UK'] # sum to 1
wealth_cols = ['Household Income Median', 'IOD AVG'] #'Well-Being'
qualilication_cols = ['No qualifications', 'Level 1 qualifications', 'Level 2 qualifications',
       'Apprenticeship qualifications', 'Level 3 qualifications',
       'Level >=4 qualifications', 'Other qualifications'] # sum to 1 
disability_cols = ['Bad Health', 'Limited activities']
all_features = age_cols + ethnicities_cols + religion_cols + born_cols + wealth_cols + qualilication_cols + disability_cols

In [51]:
def normalize(df, cols):
    total = df[cols].sum(axis=1)
    df[cols] = df[cols].div(total, axis=0)
    return df

def standartize(df, cols):
    means = df[cols].mean()
    stds = df[cols].std()
    df[cols] = (df[cols] - means) / stds
    return df

def min_max_scaling(df, cols):
    mins = df[cols].min()
    maxs = df[cols].max()
    df[cols] = (df[cols] - mins) / (maxs - mins)
    return df

In [52]:
#normalize
df_atlas[age_cols] /= 100
df_atlas = normalize(df_atlas, ethnicities_cols)
df_atlas = normalize(df_atlas, religion_cols)
df_atlas = normalize(df_atlas, born_cols)
df_atlas[qualilication_cols] /= 100

In [53]:
df_atlas['Household Income Median'] = df_atlas['Household Income Median'].apply(np.log)
df_atlas = standartize(df_atlas, wealth_cols)

## Tesco groceries preprocessing

In [54]:
important_cols = ['weight', 'volume', 'fat', 'saturate', 'salt', 'sugar', 'protein', 'carb',
                 'fibre', 'alcohol', 'energy_fat', 'energy_saturate', 'energy_sugar', 'energy_protein',
                 'energy_carb', 'energy_fibre', 'energy_alcohol', 'energy_tot', 'f_energy_fat', 'f_energy_saturate', 
                 'f_energy_sugar', 'f_energy_protein', 'f_energy_carb', 'f_energy_fibre', 'f_energy_alcohol', 
                 'energy_density', 'h_nutrients_weight', 'h_nutrients_weight_norm', 'h_nutrients_calories',
                  'h_nutrients_calories_norm', 'representativeness_norm']

In [55]:
df_atlas.set_index('New Code', drop=True, inplace=True)

In [56]:
# merge the data sets by ward
df_ward_atlas = pd.merge(df_ward[important_cols], df_atlas, how='inner', left_index=True, right_index=True, validate='1:1')

## Save dataset

In [58]:
df_ward_atlas.head(1)

weight      volume       fat  saturate     salt      sugar  \
E05000026  449.535137  125.960833  9.488797  3.693721  0.58324  10.966213   

           protein       carb     fibre   alcohol  ...  \
E05000026  4.97756  19.381951  1.564721  0.198172  ...   

           Level 2 qualifications  Apprenticeship qualifications  \
E05000026                   0.115                          0.011   

           Level 3 qualifications  Level >=4 qualifications  \
E05000026                   0.085                     0.345   

           Other qualifications  \
E05000026                 0.167   

           % of 16+ who are schoolchildren and full-time students: Age 18 and over  \
E05000026                                               14.3                         

           Bad Health  Limited activities  Well-Being   IOD AVG  
E05000026         4.3                 5.1    7.897959  0.694138  

[1 rows x 66 columns]

In [61]:
PREPROCESSED_PATH = 'preprocessed'
DATASET_NAME='df_ward_atlas.csv'
DATASET_PATH = join(FOLDER_PATH, PREPROCESSED_PATH, DATASET_NAME)
df_ward_atlas.to_csv(DATASET_PATH)